ratelimit/core.py

```diff
@@ -11,6 +11,11 @@ async def default_429(scope: Scope, receive: Receive, send: Send) -> None:
     await send({"type": "http.response.body", "body": b"", "more_body": False})
 
 
+async def default_401(scope: Scope, receive: Receive, send: Send, exc: Exception) -> None:
```

Comment:
Would it be better to use the following function?

```python
async def default_401(exc: Exception) -> ASGIApp:
    ...
```

Just like the question we discussed earlier. We can use it like this.

```python
async def handle_auth_error(exc: Exception) -> ASGIApp:
    return JSONResponse({"message": "..."}, status_code=401)
```

Reply:
Sure. I was just following `on_blocked`'s example. So what would we return in the new function?
```python
async def default_401(exc: Exception) -> ASGIApp:
    return ?
```
Why not use the same style used in the `on_blocked`  function?

Reply:
I think, should throw the exception as is in the default function. It is normal for the limiter to return 429. But if it returns something else, it doesn't seem like much of ratelimit middlware should do. It's better to let users decide for themselves.

Reply:
> Why not use the same style used in the on_blocked function?

on_blocked is a standard ASGI app. But in this function, we need to accept additional exception parameters. In order to make it easier to use other ASGI app like fastapi.JSONResponse, we should not break this interface.

Reply:
That makes sense. So since now `default_401` does nothing, should we change its default value to None and then do something like this when handling the exception?
```python
try:
    user, group = await self.authenticate(scope)
except Exception as exc:
    if self.on_auth_error is not None:
        reponse = await self.on_auth_error(exc)
        return await reponse(scope, receive, send)
    raise exc
```

Reply:
Yes, I think this is feasible.